### DISCUSSION TITLE
Can something be done about tiny-lru in webpack builds?

### DISCUSSION THREAD

**Discussion by xzilja** [fastify/fastify#2907] [General]
I recently stumbled upon this issue that is still affecting some serverless providers and webpack builds
https://github.com/fastify/fastify/issues/2356#issuecomment-675202766

Gist of it is that error of `lru is not a function` is thrown when app is built with webpack, unless following workaround is applied:

```js
  resolve: {
    alias: {
      'tiny-lru': path.join(__dirname, '../../node_modules/tiny-lru/lib/tiny-lru.cjs.js')
    },
  }
```

This is very mild annoyance, but fastify feels very "clean" in all other aspects, so I was wondering if something can be done in order to no longer require this workaround in webpack?

**Comment by Eomm** [comment]
Do you know which configuration of webpack trigger this error?

Could it be that the `tiny-lru.esm.js` is loaded cause of this?

https://github.com/avoidwork/tiny-lru/blob/683f030c52af26041af3be3f42f5a99316651197/package.json#L22

**Reply by xzilja** [reply]
Not sure which configuration, but esm module is indeed loaded by default,
This is my full config for webpack, targeting `es2020` in node 15.11

```js
const path = require('path');
const { ESBuildPlugin } = require('esbuild-loader');

module.exports = {
  entry: {
    server: path.resolve(__dirname, '../server/src/index.ts')
  },
  output: {
    filename: '[name].js',
    path: path.resolve(__dirname, '../server/dist')
  },
  resolve: {
    alias: {
      'tiny-lru': path.join(__dirname, '../../node_modules/tiny-lru/lib/tiny-lru.cjs.js')
    },
    extensions: ['.ts', '.js', '.mjs']
  },
  mode: 'production',
    optimization: {
      nodeEnv: 'production',
      splitChunks: {
        chunks: 'all'
      },
      minimize: true,
      minimizer: [
        new ESBuildMinifyPlugin({
          target: 'es2020',
          minifyWhitespace: true,
          minifyIdentifiers: true,
          minifySyntax: true
        })
      ]
  },
  target: 'node',
  module: {
    rules: [
      {
        test: /\.(m?js|ts)$/u,
        exclude: /(node_modules)$/u,
        use: {
          loader: 'esbuild-loader',
          options: {
            loader: 'ts',
            target: 'es2020',
            tsconfigRaw: require('../server/tsconfig.json')
          }
        }
      }
    ]
  },
  plugins: [new ESBuildPlugin()]
};
```

**Comment by mcollina** [comment]
I am not really sure what we can do to better support this use case. tiny-lru is one of the fastest caching libraries, and I'd be against replacing it for this.

Anyway, what error is this producing? I'd be good to land a change to support this if it's simple.

**Reply by xzilja** [reply]
Yeh, I understand, Error produced is

```zsh
TypeError: lru is not a function
   at new ContentTypeParser
   at fastify
```

**Reply by mcollina** [reply]
can you try adding a .default at the end of that require?

**Reply by xzilja** [reply]
For lru? I'm not importing / using it myself, as for fastify, I use `import fastify from "fastify"` as I am in typescript environment.

**Reply by xzilja** [reply]
Ah, so if I add `.default` to its require in fastify's source https://github.com/fastify/fastify/blob/fefab178bc1cc14cfd5965583500800ff0c61d27/lib/contentTypeParser.js#L3

It works 👍 

Is this a possible change, or will it break other environments?

**Reply by mcollina** [reply]
We'll need to do this with an if. Could you send a PR?

**Reply by xzilja** [reply]
Can do, what will `if` need to check for however?

```js
if (require('tiny-lru').default) {
 // ...?
}
```

**Reply by mcollina** [reply]
```
let LRU = require('tiny-lru')
LRU = typeof LRU === 'function' ? LRU : LRU.default
```
